fdbkubernetesmonitor/monitor.go

```diff
@@ -0,0 +1,346 @@
+// monitor.go
+//
+// This source file is part of the FoundationDB open source project
+//
+// Copyright 2021 Apple Inc. and the FoundationDB project authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+)
+
+// errorBackoffSeconds is the time to wait after a process fails before starting
+// another process.
+// This delay will only be applied when there has been more than one failure
+// within this time window.
+const errorBackoffSeconds = 60
+
+// Monitor provides the main monitor loop
+type Monitor struct {
+	// ConfigFile defines the path to the config file to load.
+	ConfigFile string
+
+	// CustomEnvironment defines the custom environment variables to use when
+	// interpreting the monitor configuration.
+	CustomEnvironment map[string]string
+
+	// ActiveConfiguration defines the active process configuration.
+	ActiveConfiguration *ProcessConfiguration
+
+	// ActiveConfigurationBytes defines the source data for the active process
+	// configuration.
+	ActiveConfigurationBytes []byte
+
+	// LastConfigurationTime is the last time we successfully reloaded the
+	// configuration file.
+	LastConfigurationTime time.Time
+
+	// ProcessIDs stores the PIDs of the processes that are running. A PID of
+	// zero will indicate that a process does not have a run loop. A PID of -1
+	// will indicate that a process has a run loop but is not currently running
+	// the subprocess.
+	ProcessIDs []int
+
+	// Mutex defines a mutex around working with configuration.
+	Mutex sync.Mutex
```

Comment:
I'm not sure what this Mutex is protecting, or when it needs to be held.  For instance, it's not held when you call LoadConfiguration (unless I'm missing something), and there's no note explaining why.

The multithreaded invariants would be clearer if you pulled all accesses to `ProcessIDs` and `configuration.ServerCount` into private methods that started with the lines:
```
mutex.Lock()
defer monitor.mutex.Unlock()
```
Ideally, they'd be in their own sub-object so it's mutex locking protocol was encapsulated away from the lock-free parts of this data structure.  That structure's documentation could explain the locking protocol.